core/src/layers/mime_guess.rs

```diff
@@ -0,0 +1,324 @@
+/// - [Operator::stat](../struct.Operator.html#method.stat)
+/// - [Operator::stat_with](../struct.Operator.html#method.stat_with)
+/// - [Operator::list_with](../struct.Operator.html#method.list_with)
+/// - [Operator::lister_with](../struct.Operator.html#method.lister_with)
+/// - [BlockingOperator::write](../struct.BlockingOperator.html#method.write)
+/// - [BlockingOperator::write_with](../struct.BlockingOperator.html#method.write_with)
+/// - [BlockingOperator::writer](../struct.BlockingOperator.html#method.writer)
+/// - [BlockingOperator::writer_with](../struct.BlockingOperator.html#method.writer_with)
+/// - [BlockingOperator::stat](../struct.BlockingOperator.html#method.stat)
+/// - [BlockingOperator::stat_with](../struct.BlockingOperator.html#method.stat_with)
+/// - [BlockingOperator::list_with](../struct.BlockingOperator.html#method.list_with)
+/// - [BlockingOperator::lister_with](../struct.BlockingOperator.html#method.lister_with)
+///
+/// Your operation will automatically carry `Content-Type` information.
+///
+/// However, please note that this layer will not overwrite the `content_type` you manually set,
+/// nor will it overwrite the `content_type` provided by backend services.
+///
+/// A simple example is that for object storage backends, when you call `stat`, the backend will
+/// provide `content_type` information, and `mime_guess` will not be called, but will use
+/// the `content_type` provided by the backend.
+///
+/// But if you use the [Fs](../services/struct.Fs.html) backend to call `stat`, the backend will
+/// not provide `content_type` information, and our `mime_guess` will be called to provide you with
+/// appropriate `content_type` information.
+///
+/// Another thing to note is that using this layer does not necessarily mean that the result will 100%
+/// contain `content_type` information. If the extension of your path is custom or an uncommon type,
+/// the returned result will still not contain `content_type` information (the specific condition here is
+/// when [mime_guess::from_path::first_raw](https://docs.rs/mime_guess/latest/mime_guess/struct.MimeGuess.html#method.first_raw)
+/// returns `None`).
+///
+/// # Examples
+///
+/// ```no_run
+/// use anyhow::Result;
+/// use opendal::layers::MimeGuessLayer;
+/// use opendal::services;
+/// use opendal::Operator;
+/// use opendal::Scheme;
+///
+/// let _ = Operator::new(services::Memory::default())
+///     .expect("must init")
+///     .layer(MimeGuessLayer::default())
+///     .finish();
+/// ```
+#[derive(Debug, Copy, Clone, Default)]
+// Developer note:
+// The inclusion of a private unit tuple inside the struct here is to force users to
+// use `MimeGuessLayer::default()` instead of directly using `MimeGuessLayer` to
+// construct instances.
+// This way, when we add some optional config methods to this layer in the future,
+// the old code can still work perfectly without any breaking changes.
+pub struct MimeGuessLayer(());
```

Comment:
Let's use `MimeGuessLayer {}` instead.

Reply:
How to use only `MimeGuessLayer {}` so that users can only construct it through `MimeGuessLayer::default()`? In this way, users can still construct directly through the `MimeGuessLayer {}`. Should something else be placed inside the structure?

Reply:
We can add a [`#[non_exhaustive]`](https://doc.rust-lang.org/reference/attributes/type_system.html#the-non_exhaustive-attribute) here to prevent users to build this layer by `MimeGuessLayer {}`.

Reply:
Thank! Got a new trick! 😎